lib/rspec/mocks/method_double.rb

```diff
@@ -249,6 +249,11 @@ def new_rspec_prepended_module
         end
       end
 
+      def method_owner
+        @method_owner ||=
```

Comment:
Does it affect the performance?

Reply:
It's memoized so I don't think we'll see anything noticeable in practice. If we disregard memoization and run simple benchmark, `object.method(method_name).owner` is about 3 times faster:

```
Warming up --------------------------------------
 object.method.owner   261.517k i/100ms
       bind and call    91.219k i/100ms
Calculating -------------------------------------
 object.method.owner      2.661M (± 0.9%) i/s -     13.337M in   5.011938s
       bind and call    914.184k (± 1.0%) i/s -      4.652M in   5.089434s
```

From running this:

``` ruby

require 'benchmark/ips'

Benchmark.ips do |x|
  object = Object.new

  x.report('object.method.owner') do
    object.method(:to_s).owner
  end

  x.report('bind and call') do
    Object.instance_method(:method).bind(object).call(:to_s).owner
  end
end
```

Well, or we can remove the test and calling out that do not override `method`, which I fully support :P

Reply:
I'm clearly on your side with you that one shouldn't override `Kernel`'s internals with unrelated application-level implementations. However, someone's code may break with a weird message, and they will open a ticket, and we'll have to address it, since even though this might not be a good style to override `method`, it's not our call to enforce that in any way.

Reply:
Yeah, of course, that's why I was half joking there :)

Half joking, half serious, because there are a lot of other methods we're relying on, and of course we won't want to be so defensive to do it this way for most of the built-in methods? (or a rule of thumb is, whenever someone is filing a ticket 🤷 This is also relying on someone isn't trying to break RSpec intentionally)

I see `method` or `send` is more commonly overridden with a completely different behaviour (if it's the same behaviour like some people will override `method_missing` also with `respond_to_missing?` properly, that would be mostly alright, but issues might still happen especially for mocks implementation which often talks about proxies), and that's why we have `__send__` we can use, sadly not for `method` and it may be confusing that `__method__` is actually a completely different method! Not like `__send__` is an alias of `send`.

Sorry that this is probably falling down to a rant. To be more on the topic, since the test is already there, I don't think a regression is acceptable. However if we never added that test, I would think it's fine to just say "don't override `method`". Maybe RSpec 5, or maybe it doesn't matter after all. It's just one method right now. We don't have an army of `Object.instance_method(method_name).bind(object)` yet. Or do we? Maybe it's inside one of the support libraries already? 😅 To be honest I won't be surprised!

Reply:
We defend against a number of quite odd things:
```
# This class protects us against Mutex.new stubbed out within tests
```
https://github.com/rspec/rspec-core/pull/2886, https://github.com/rspec/rspec-support/pull/431 and in a few other places.

> Maybe it's inside one of the support libraries already? 😅

Good point, it is! 😄 
https://github.com/rspec/rspec-support/blob/d63133f478408c1d965e673b96ad10ef5a5d183f/lib/rspec/support.rb#L53

Reply:
> We defend against a number of quite odd things:

Oh my, I was guilty here. I am sure I was once lazy and stubbed `File.read` without explicit about the arguments, and it caused some issues. However being explicit about a specific argument did the trick for me.

I wonder if this kinds of things will need a more general approach. For example, will this work?

``` ruby
RSpec::Support::Source::File = ::File.dup
```

> Good point, it is! 😄

Ok, thank you for spotting that. I updated the code to use it instead.

Reply:
> kinds of things

I can think of three cases we've touched:
1. Ruby core method stubbed (`File.read`, `File.expand_path`, `Mutex.new`)
2. Ruby core method overridden in a subclass (`Kernel.method`)
3. Attempt to define a method with memoized helpers that would break `Example` (`to_s`, `initialize`)

Not sure if there could be a common solution to all of this.

Reply:
> Ruby core method stubbed (File.read, File.expand_path, Mutex.new)

I just tried this does work:

``` ruby
PreservedFile = File.dup

def File.read
  'break you'
end

PreservedFile.read('README.md')
```

We can have a list of preserved classes.

``` ruby
module RSpec
  module Preserved
    File = ::File.dup
    Mutex = ::Mutex.dup
  end
end
```

> Ruby core method overridden in a subclass (Kernel.method)

We can always do it via the same trick.

``` ruby
module RSpec
  module Preserved
    module_function
    def send_kernel(object, message, *args, **kargs)
      ::Kernel.instance_method(message).bind(object).call(*args, **kargs)
    end
  end
end

method_owner = RSpec::Preserved.send_kernel(object, :method).owner
```

> Attempt to define a method with memoized helpers that would break Example (to_s, initialize)

For `to_s` and `initialize` I think we should just raise an error like now, indeed. It would be too much work for those to be overridden. Even if we can stick with `__to_s__` or `__initialize__` and play around with `allocate`, it's easily broken with other integration.

> Not sure if there could be a common solution to all of this.

I didn't mean a common solution to all of above. I meant a common solution for the same issue. For example, we use the same way to resolve `File.read` or `Mutex.new` (a module method) got overridden. We use the same way to resolve a particular instance method got overridden.

Reply:
Our policy is to do the method work around for cases as they occur